flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/KafkaSourceBuilder.java

```diff
@@ -458,6 +458,29 @@ private boolean maybeOverride(String key, String value, boolean override) {
         return overridden;
     }
 
+    private boolean maybeOverridePartitionDiscovery(String key, String value, boolean override) {
```

Comment:
override->isBounded maybe more readable

Reply:
i think override is readable. “isBounded” also indicates that the property needs to be overridden

Reply:
It looks that we should only disable auto discovery here in the following two cases:
1. The source is bounded.
2. The source has stopping offsets and the user did not explicitly set the discovery interval. 

In other cases, we don't need to do anything. So the logic here could just become something like:
```
boolean hasStoppingOffsets = !(stoppingOffsetsInitializer instanceof NoStoppingOffsetsInitializer);
boolean hasParitionDiscoverySetting = props.getProperty(KafkaSourceOptions.PARTITION_DISCOVERY_INTERVAL_MS.key()) != null;
if (boundedness == Bounded || 
(hasStoppingOffsets && !hasParititionDiscoverySetting)) {
    props.set(KafkaSourceOptions.PARTITION_DISCOVERY_INTERVAL_MS.key(-1))
}
```
It seems easier to understand.

BTW, can we also add a unit test for this?

Reply:
Thanks, @becketqin .okey,I will take your suggestion and add a unit test!

Reply:
Hi,@becketqin.Does this unit test only need to check the properties of KafkaSource or partitionDiscoveryIntervalMs of KafkaSourceEnumerator?